hips/hip-9999.md

```diff
@@ -0,0 +1,72 @@
+---
+hip: 9999
+title: "Support Helm release data stored in multiple K8s Secrets"
+authors: [ "Ralf Van Dorsselaer" ]
+created: "2022-06-20"
+type: "feature"
+status: "draft"
+---
+ 
+## Abstract
+ 
+Helm release data encoded into the release field in a Kubernetes Secret can cause the Secret as a whole to exceed 1 MiB (as defined by MaxSecretSize in kubernetes/pkg/apis/core/types.go).
+ 
+When this happens, the Kubernetes Secret cannot be stored in Etcd and the Helm install or upgrade fails.
+ 
+Splitting the Helm release data across multiple Kubernetes Secrets resolves this problem.
+ 
+## Motivation
+ 
+Helm release data that causes the Kubernetes Secret size to exceed MaxSecretSize will make Helm install or upgrade fail.
+ 
+The existing HELM_DRIVER=secret storage driver does not handle Kubernetes Secrets become too big when storing a large encoded Helm release data.
+ 
+## Rationale
+ 
+The logical place to support this is in the HELM_DRIVER=secret storage driver.
+ 
+## Specification
+ 
+The feature is transparent to users. 
+ 
+When the HELM_DRIVER=secret storage driver notices that the resulting Kubernetes Secret will exceed MaxSecretSize then the storage driver will split the encoded release data across multiple Kubernetes Secrets.
+ 
+Splitting requires additional metadata to be added to the Kubernetes Secret:
+ 
+- number of chunks: integer, the number of Kubernetes Secrets that holds the release data for this Helm release version.
+- chunk: integer, the chunk number. Each Secret gets a chunk number. Reconstituting the release data can be done by reading the Secret with chunk=1, then subsequently reading in numerical order any number of additional Secrets up to and including the "number of chunks".
+ 
+Splitting requires the name of the Secret to include a "chunk" suffix. Example: sh.helm.release.v1.test.v1.2, 3, 4 and so on where the number after the last dot is the "number of chunk". The first Secret will not get a .1 suffix.
+ 
+If no splitting is required then no additional metadata is included/set.
+ 
+## Backwards compatibility
+ 
+If the chunking is implemented in the secret/secrets HELM_DRIVER then backward compatibility to previously installed Helm releases is guaranteed. If no chunking information is encoded in the release secret then the driver can behave as before chunking support was added. If no chunking is needed then the release secret can be created without any additional metadata and with compatible name.
+ 
+## Security implications
+ 
+Theoretically a MTM Man in The Middle attack is possible where requests to the Kubernetes API server are intercepted. But this type of attack applies to any access to the Kubernetes API server.
```

Comment:
etcd relies on the fact that key/values are no larger than 1MB in size. By chunking key/values into smaller sizes just so they can fit in etcd does not address the root issue: introducing more read requests will increase the latency of other read requests. It is theoretically possible that a release secret that is >1MB in size could potentially DDOS etcd. Because etcd has to dedicate a large amount of resources to read several of keys, other incoming read requests could grind to a halt.

Also worth keeping in mind that etcd has [a default hard storage limit](https://etcd.io/docs/v3.5/dev-guide/limit/#storage-size-limit) at 2GB. Even if you allow release secrets larger than 1MB, you will very quickly run out of space.

I am incredibly wary of "working around" a limitation that was explicitly put in place by the etcd engineers.

Reply:
True, but IMO only in theory. The idea is not to make the chunks smaller but to distribute a Helm release that comes out to 1.5MB across two Secrets one being 1MB the other 512K. In implementation I would not make the chunksize configurable in any way. (It is for testing in the current PR)